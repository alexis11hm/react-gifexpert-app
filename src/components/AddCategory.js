import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {


    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        setInputValue(e.target.value)
    }

    const handleSubmit = (e) => {
        e.preventDefault()

        if(inputValue.trim().length > 2){
            setCategories(categories => [inputValue,...categories])
            setInputValue('')
        }
    }

    return (
        //si tenemos un elemento que agrupe a otros elementos no es necesario el fragment <></> en este caso, form
        <form onSubmit={handleSubmit}>
                <input 
                    type="text"
                    value={inputValue}
                    onChange={handleInputChange}
                />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
