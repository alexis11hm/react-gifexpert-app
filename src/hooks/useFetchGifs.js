import { useState, useEffect } from "react"

import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data:[],
        loading:true
    })

    //Solo se dispara una vez cuando el componente se renderiza por primera vez
    //se hace una validacion en los [] para saber cuando se renderiza de nuevo
    //los efectos no pueden ser asincronos
    useEffect( () => {
        getGifs(category).then(imgs => {
            setState({
                    data: imgs,
                    loading: false
                })
            }
        )
        
    }, [category]) //Si la categoria cambia, entonces se volvera a ejecutar el useEffect
    

    return state //{ data: [], loading: true }
}