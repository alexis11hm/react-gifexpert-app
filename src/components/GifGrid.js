import React/*, { useState, useEffect }*/ from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem'

import PropTypes from 'prop-types'

export const GifGrid = ({category}) => {

    //data cambian al nombre de images, pero el contenido es lo mismo
    const {data:images, loading} = useFetchGifs(category)

    return (
        <>
            <h3>{category}</h3>
            {loading && <p className="animate__animated animate__flash">Loading</p>}
            <div className="card-grid">
                    {
                        //{..img} es como pasar los atributos de img [id,title, url] y su valor como props, pero de manera corta
                        images.map( (img) => (
                            <GifGridItem 
                                key={ img.id } 
                                { ...img }
                            /> 
                            )
                        )
                    }
            </div>
        </>
    )
}

GifGrid.propTypes ={
    category: PropTypes.string.isRequired
}
