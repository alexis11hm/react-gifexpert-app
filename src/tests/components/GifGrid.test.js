import { shallow } from "enzyme"
import { GifGrid } from "../../components/GifGrid"
import { useFetchGifs } from "../../hooks/useFetchGifs"

import '@testing-library/jest-dom'

//fingir cualquier llamada al archivo y controlar la informacion que respondra
jest.mock('../../hooks/useFetchGifs')

describe('Pruebas en el <GifGrid />',()=>{

    const category = 'One punch'

    test('debe de mostrarse correctamente',()=>{

        //cuando se llama el hook en el componente regresa el valor
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        })

        const wrapper = shallow(<GifGrid category={category} /> )
        expect(wrapper).toMatchSnapshot()
    })

    test('debe de mostrar items cuando se cargan imagenes con useFetchGifs',() => {
        
        const gifs = [
        {
            id:'abc',
            title:'titulo',
            url:'https://localhost/any.png'
        },
        {
            id:'def',
            title:'titulo',
            url:'https://localhost/any.png'
        }
    ]

        //cuando se llama el hook en el componente regresa el valor
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        })

        const wrapper = shallow(<GifGrid category={category} /> )
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('p').exists()).toBe(false)
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
    })

})