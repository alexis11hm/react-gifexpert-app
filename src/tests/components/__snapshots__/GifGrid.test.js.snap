// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <GifGrid /> debe de mostrar items cuando se cargan imagenes con useFetchGifs 1`] = `
<Fragment>
  <h3>
    One punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="abc"
      key="abc"
      title="titulo"
      url="https://localhost/any.png"
    />
    <GifGridItem
      id="def"
      key="def"
      title="titulo"
      url="https://localhost/any.png"
    />
  </div>
</Fragment>
`;

exports[`Pruebas en el <GifGrid /> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3>
    One punch
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
